scipy/stats/_stats_py.py

```diff
@@ -100,13 +100,20 @@ def _contains_nan(a, nan_policy='propagate'):
         try:
             contains_nan = np.any(np.isnan(a))
         except TypeError:
-            # Don't know what to do. Fall back to omitting nan values and
-            # issue a warning.
-            contains_nan = False
-            nan_policy = 'omit'
-            warnings.warn("The input array could not be properly "
-                          "checked for nan values. nan values "
-                          "will be ignored.", RuntimeWarning)
+            try:
+                # This can happen when attempting to check nan with np.isnan
+                # for string array (e.g. as in the function `rankdata`).
+                a = np.array(a)
+                nan_str = np.array([np.nan]).astype(str)[0]
+                contains_nan = a[np.where(a.astype(str) == nan_str)].size != 0
```

Comment:
I was thrown off when I first saw this because it looked very complex. 

This also works:
```suggestion
                contains_nan = np.any(np.asarray(a) == "nan")
```

And the only reason `np.asarray(a)` is needed here is because the tests pass in lists, but in practice `a` is always an array before being passed into `_contains_nan`. (All functions that call `_contains_nan` call one of `asarray`, `asanyarray`, or `_chk_asarray` before calling `_contains_nan`.)

So can we justify the extra complexity of what is here? If so, we can keep it, but we should add a comment in the code to explain why it's necessary. I'll consider each line separately.

Reply:
> it looked very complex.
Agree. I will try update this PR.